File: server/controllers/lastMinuteController.js
const { ProfileRepository, HistoryRepository } = require('../models/dbStore');
const llmGateway = require('../services/llmGateway');

// @desc    Generate compact last-minute interview revision guide
// @route   POST /api/last-minute/generate
const generateLastMinuteGuide = async (req, res, next) => {
  try {
    let { company, role, jobDescription, resumeText } = req.body;

    const profile = await ProfileRepository.findOne({ userId: req.user._id });
    if (profile) {
      if (!company) company = profile.company;
      if (!role) role = profile.role;
      if (!jobDescription) jobDescription = profile.jobDescription;
      if (!resumeText) resumeText = profile.resumeText;
    }

    const targetCompany = company || 'Target Company';
    const targetRole = role || 'Software Engineer';

    const prompt = `
Generate a compact, highly scannable Last-Minute Interview Revision Guide for a candidate interviewing for "${targetRole}" at "${targetCompany}".
Include:
1. "topTechnicalTopics": 4 bullet points on the most crucial concepts to review right now.
2. "mustKnowQuestions": 3 vital Q&A pairs that are almost guaranteed to be asked.
3. "quickChecklist": 4 actionable pre-interview sanity check items.
4. "smartQuestionsToAsk": 3 insightful questions the candidate can ask the interviewer.

Return ONLY valid JSON with this schema:
{
  "topTechnicalTopics": ["...", "..."],
  "mustKnowQuestions": [
    { "q": "...", "a": "..." }
  ],
  "quickChecklist": ["...", "..."],
  "smartQuestionsToAsk": ["...", "..."]
}
`;

    const context = {
      company: targetCompany,
      role: targetRole,
      jobDescription,
      resumeText,
    };

    const aiResult = await llmGateway.generateAIResponse({
      prompt,
      context,
      jsonMode: true,
      moduleType: 'last-minute',
      systemPrompt: 'You are a Senior Interview Coach delivering a high-yield pre-interview survival sheet.',
    });

    const data = aiResult.parsedJson || llmGateway.generateFallbackResponse('last-minute', context);

    await HistoryRepository.create({
      userId: req.user._id,
      module: 'Last Minute Guide',
      company: targetCompany,
      role: targetRole,
      summary: `Generated last-minute revision checklist for ${targetCompany}`,
      metadata: { targetCompany },
    });

    res.status(200).json({
      success: true,
      data,
      provider: aiResult.provider,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  generateLastMinuteGuide,
};

File: server/src/main/java/com/nexoffer/controller/LastMinuteController.java
package com.nexoffer.controller;

import com.nexoffer.model.History;
import com.nexoffer.model.Profile;
import com.nexoffer.model.User;
import com.nexoffer.repository.HistoryRepository;
import com.nexoffer.repository.ProfileRepository;
import com.nexoffer.service.LlmGatewayService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/last-minute")
public class LastMinuteController {

    private final ProfileRepository profileRepository;
    private final HistoryRepository historyRepository;
    private final LlmGatewayService llmGatewayService;

    public LastMinuteController(ProfileRepository profileRepository,
                                HistoryRepository historyRepository,
                                LlmGatewayService llmGatewayService) {
        this.profileRepository = profileRepository;
        this.historyRepository = historyRepository;
        this.llmGatewayService = llmGatewayService;
    }

    @PostMapping("/generate")
    public ResponseEntity<?> generateLastMinuteGuide(@AuthenticationPrincipal User user, @RequestBody(required = false) Map<String, String> body) {
        String company = body != null ? body.get("company") : null;
        String role = body != null ? body.get("role") : null;
        String jobDescription = body != null ? body.get("jobDescription") : null;
        String resumeText = body != null ? body.get("resumeText") : null;

        Optional<Profile> profileOpt = profileRepository.findByUserId(user.getId());
        if (profileOpt.isPresent()) {
            Profile p = profileOpt.get();
            if (company == null || company.isBlank()) company = p.getCompany();
            if (role == null || role.isBlank()) role = p.getRole();
            if (jobDescription == null || jobDescription.isBlank()) jobDescription = p.getJobDescription();
            if (resumeText == null || resumeText.isBlank()) resumeText = p.getResumeText();
        }

        String targetCompany = (company != null && !company.isBlank()) ? company : "Target Company";
        String targetRole = (role != null && !role.isBlank()) ? role : "Software Engineer";

        String prompt = "Generate a compact, highly scannable Last-Minute Interview Revision Guide for a candidate interviewing for \"" + targetRole + "\" at \"" + targetCompany + "\".\n\n" +
                "Return ONLY valid JSON with this schema:\n" +
                "{\n" +
                "  \"topTechnicalTopics\": [\"...\"],\n" +
                "  \"mustKnowQuestions\": [{ \"q\": \"...\", \"a\": \"...\" }],\n" +
                "  \"quickChecklist\": [\"...\"],\n" +
                "  \"smartQuestionsToAsk\": [\"...\"]\n" +
                "}";

        Map<String, String> context = new HashMap<>();
        context.put("company", targetCompany);
        context.put("role", targetRole);
        if (jobDescription != null) context.put("jobDescription", jobDescription);
        if (resumeText != null) context.put("resumeText", resumeText);

        LlmGatewayService.AiResponse aiResponse = llmGatewayService.generateAIResponse(
                prompt,
                "You are a Senior Interview Coach delivering a high-yield pre-interview survival sheet.",
                context,
                true,
                "last-minute"
        );

        Object data = aiResponse.getParsedJson() != null ? aiResponse.getParsedJson() : llmGatewayService.generateFallbackResponse("last-minute", context);

        historyRepository.save(new History(
                user.getId(),
                "Last Minute Guide",
                targetCompany,
                targetRole,
                "Generated pre-interview survival cheat sheet"
        ));

        return ResponseEntity.ok(Map.of(
                "success", true,
                "data", data,
                "provider", aiResponse.getProvider()
        ));
    }
}
